### ISSUE TITLE
Extend Rules Api to support register/unregister a set of rules

### ISSUE THREAD

**Issue by Desislav-Petrov** [j-easy/easy-rules#288]
The current API in Rules supports register/deregister a single rule. In some cases, however, you might want to register a whole list of already instantiated rules.

The current API wouldn't gracefully support doing that and you would have to wrap it into a while loop.

My proposal is to add an overloaded pair of methods as follows:

register(List<Object> rules)
unregister(List<Object> rules)

I'm happy to do a PR for that if we think this is the right way forward.

Thanks,
Des

**Comment by benas** [comment]
Do we really need this? We can already create a `Rules` object by passing a [vararg of rules](https://github.com/j-easy/easy-rules/blob/master/easy-rules-core/src/main/java/org/jeasy/rules/api/Rules.java#L58):

```java
// given a list of rules
List<Rule> ruleList = Arrays.asList(rule1, rule2);
// register all rules
Rules rules = new Rules(ruleList.toArray());
```

We can also register them after creating the `Rules` object if needed:

```java
// given a list of rules
List<Rule> ruleList = Arrays.asList(rule1, rule2);
Rules rules = new Rules();
// register all rules
ruleList.forEach(rules::register);
```

Do you agree?

**Comment by Desislav-Petrov** [comment]
Hi,

Thanks for your comments.

The way I was thinking about it is that the Rules class is essentially a Collection. As such, it makes sense to have the add,remove,addAll,removeAll & size methods. 

I have spotted that the Rules class implements just the Iterable interface so based on this I'd agree with you. However, I'm also interested what was the motivation for the Rules class to implement just Iterable rather than Collection iface? 

Regards,
Des

**Comment by benas** [comment]
There were no special motivations for that. The goal was to make this abstraction iterable and act as a namespace for a rule set, hence the naming register/unregister instead of add/remove. It is not incorrect to see it as a collection, specifically a Set since it can't contain duplicate rules.

I prefer a single method with a vararg argument instead of having two separate methods to register/unregister a single element and a list of elements. What about making the current register/unregister methods accept a vararg of rules? Something like:

```diff
--public void register(Object rule) {
--    Objects.requireNonNull(rule);
--    rules.add(RuleProxy.asRule(rule));
--}
++public void register(Object... rules) {
++    for (Object rule : rules) {
++        Objects.requireNonNull(rule);
++        this.rules.add(RuleProxy.asRule(rule));
++    }
++}
```

Same for `unregister`. This would be a non-breaking change that we can introduce in the upcoming v4.1. If you agree, you can update the PR accordingly and it should be good to merge.

Let me know about your thoughts.

**Comment by Desislav-Petrov** [comment]
Hi,

I like this idea and is backwards compatible. I'll update the PR over the weekend and we can merge.

Regards,
Des

**Comment by benas** [comment]
Resolved with #289 .
